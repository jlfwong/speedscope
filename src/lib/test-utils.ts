import * as fs from 'fs'
import * as path from 'path'
import {Profile, CallTreeNode, Frame} from './profile'
import {importProfileGroup} from '../import'
import {exportProfileGroup, importSpeedscopeProfiles} from './file-format'

interface DumpedProfile {
  name: string
  stacks: string[]
  frames: Frame[]
}

export function dumpProfile(profile: Profile): any {
  const dump: DumpedProfile = {
    name: profile.getName(),
    stacks: [],
    frames: [],
  }

  profile.forEachFrame(f => dump.frames.push(f))

  let lastValue = 0
  const curStack: (number | string)[] = []

  function maybeEmit(value: number) {
    if (lastValue != value) {
      dump.stacks.push(
        curStack.map(k => `${k}`).join(';') + ` ${profile.formatValue(value - lastValue)}`,
      )
      lastValue = value
    }
  }

  function openFrame(node: CallTreeNode, value: number) {
    maybeEmit(value)
    curStack.push(node.frame.name)
  }

  function closeFrame(node: CallTreeNode, value: number) {
    maybeEmit(value)
    curStack.pop()
  }

  profile.forEachCall(openFrame, closeFrame)

  return dump
}

export async function checkProfileSnapshot(filepath: string) {
  const input = fs.readFileSync(filepath, 'utf8')

  const profileGroup = await importProfileGroup(path.basename(filepath), input)
  if (profileGroup) {
    expect(profileGroup.name).toMatchSnapshot('profileGroup.name')
    expect(profileGroup.indexToView).toMatchSnapshot('indexToView')

    for (let profile of profileGroup.profiles) {
      expect(dumpProfile(profile)).toMatchSnapshot()
    }
  } else {
    fail('Failed to extract profile')
    return
  }

  const profilesWithoutFilename = await importProfileGroup('unknown', input)
  if (profilesWithoutFilename) {
    expect(profilesWithoutFilename.profiles.length).toEqual(profileGroup.profiles.length)
    profilesWithoutFilename.name = profileGroup.name
    for (let i = 0; i < profileGroup.profiles.length; i++) {
      const a = profileGroup.profiles[i]
      const b = profilesWithoutFilename.profiles[i]
      b.setName(a.getName())
    }
    expect(exportProfileGroup(profileGroup)).toEqual(exportProfileGroup(profilesWithoutFilename))
  } else {
    fail('Failed to extract profile when filename was "unknown"')
    return
  }

  const exported = exportProfileGroup(profileGroup)
  const reimportedGroup = importSpeedscopeProfiles(exported)

  expect(reimportedGroup.name).toEqual(profileGroup.name)
  expect(reimportedGroup.profiles.length).toEqual(profileGroup.profiles.length)

  for (let i = 0; i < profileGroup.profiles.length; i++) {
    const profile = profileGroup.profiles[i]
    const reimported = reimportedGroup.profiles[i]

    expect(reimported.getTotalWeight()).toEqual(profile.getTotalWeight())
    expect(dumpProfile(reimported).stacks.join('\n')).toEqual(
      dumpProfile(profile).stacks.join('\n'),
    )
  }

  const reexported = exportProfileGroup(reimportedGroup)
  expect(exported).toEqual(reexported)
}
