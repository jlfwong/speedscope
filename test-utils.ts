import * as fs from 'fs'
import * as path from 'path'
import {Profile, CallTreeNode, Frame} from './profile'
import {importProfile} from './import'

interface DumpedProfile {
  stacks: string[]
  frames: Frame[]
}

export function dumpProfile(profile: Profile): any {
  const dump: DumpedProfile = {
    stacks: [],
    frames: [],
  }

  profile.forEachFrame(f => dump.frames.push(f))

  let lastValue = 0
  const curStack: (number | string)[] = []

  function maybeEmit(value: number) {
    if (lastValue != value) {
      dump.stacks.push(
        curStack.map(k => `${k}`).join(';') + ` ${profile.formatValue(value - lastValue)}`,
      )
      lastValue = value
    }
  }

  function openFrame(node: CallTreeNode, value: number) {
    maybeEmit(value)
    curStack.push(node.frame.name)
  }

  function closeFrame(node: CallTreeNode, value: number) {
    maybeEmit(value)
    curStack.pop()
  }

  profile.forEachCall(openFrame, closeFrame)

  return dump
}

export async function checkProfileSnapshot(filepath: string) {
  const input = fs.readFileSync(filepath, 'utf8')
  const profile = await importProfile(path.basename(filepath), input)
  if (profile) {
    expect(dumpProfile(profile)).toMatchSnapshot()
  } else {
    fail('Failed to extract profile')
  }
}
