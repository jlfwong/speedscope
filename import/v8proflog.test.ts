import * as child_process from 'child_process'
import {checkProfileSnapshot} from '../test-utils'

test('importFromV8ProfLog', async () => {
  const isolatePath = './sample/profiles/node/isolate-0x102802600-v8.log'
  const v8logJsonPath = './sample/profiles/node/simple.v8log.json'
  child_process.execSync(
    `${process.execPath} --prof-process --preprocess -j ${isolatePath} > ${v8logJsonPath}`,
  )
  await checkProfileSnapshot(v8logJsonPath)
})
