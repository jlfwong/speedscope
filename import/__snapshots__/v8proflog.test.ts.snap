// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`importFromV8ProfLog 1`] = `
Object {
  "frames": Array [
    Frame {
      "col": NaN,
      "file": "bootstrap_node.js",
      "key": " bootstrap_node.js:10:10",
      "line": 10,
      "name": "(anonymous)",
      "selfWeight": 0,
      "totalWeight": 116424,
    },
    Frame {
      "col": NaN,
      "file": "bootstrap_node.js",
      "key": "startup bootstrap_node.js:12:19",
      "line": 12,
      "name": "startup",
      "selfWeight": 0,
      "totalWeight": 116424,
    },
    Frame {
      "col": NaN,
      "file": "bootstrap_node.js",
      "key": "setupGlobalVariables bootstrap_node.js:265:32",
      "line": 265,
      "name": "setupGlobalVariables",
      "selfWeight": 0,
      "totalWeight": 29385,
    },
    Frame {
      "col": NaN,
      "file": "bootstrap_node.js",
      "key": "NativeModule.require bootstrap_node.js:534:34",
      "line": 534,
      "name": "NativeModule.require",
      "selfWeight": 0,
      "totalWeight": 52273,
    },
    Frame {
      "col": NaN,
      "file": "bootstrap_node.js",
      "key": "NativeModule.compile bootstrap_node.js:602:44",
      "line": 602,
      "name": "NativeModule.compile",
      "selfWeight": 0,
      "totalWeight": 52273,
    },
    Frame {
      "col": NaN,
      "file": "util.js",
      "key": " util.js:1:11",
      "line": 1,
      "name": "(anonymous)",
      "selfWeight": 0,
      "totalWeight": 29385,
    },
    Frame {
      "col": NaN,
      "file": "internal/encoding.js",
      "key": " internal/encoding.js:1:11",
      "line": 1,
      "name": "(anonymous)",
      "selfWeight": 0,
      "totalWeight": 29385,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": "(c++) v8::internal::Runtime_CreateArrayLiteral",
      "line": undefined,
      "name": "(c++) v8::internal::Runtime_CreateArrayLiteral",
      "selfWeight": 0,
      "totalWeight": 29385,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": "(c++) v8::internal::JSFunction::EnsureHasInitialMap",
      "line": undefined,
      "name": "(c++) v8::internal::JSFunction::EnsureHasInitialMap",
      "selfWeight": 29385,
      "totalWeight": 29385,
    },
    Frame {
      "col": NaN,
      "file": "bootstrap_node.js",
      "key": "setupGlobalConsole bootstrap_node.js:320:30",
      "line": 320,
      "name": "setupGlobalConsole",
      "selfWeight": 0,
      "totalWeight": 22888,
    },
    Frame {
      "col": NaN,
      "file": "bootstrap_node.js",
      "key": "setupInspectorCommandLineAPI bootstrap_node.js:364:40",
      "line": 364,
      "name": "setupInspectorCommandLineAPI",
      "selfWeight": 0,
      "totalWeight": 22888,
    },
    Frame {
      "col": NaN,
      "file": "module.js",
      "key": " module.js:1:11",
      "line": 1,
      "name": "(anonymous)",
      "selfWeight": 0,
      "totalWeight": 22888,
    },
    Frame {
      "col": NaN,
      "file": "fs.js",
      "key": " fs.js:1:11",
      "line": 1,
      "name": "(anonymous)",
      "selfWeight": 0,
      "totalWeight": 22888,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": "(c++) v8::internal::Runtime_StoreIC_Miss",
      "line": undefined,
      "name": "(c++) v8::internal::Runtime_StoreIC_Miss",
      "selfWeight": 0,
      "totalWeight": 22888,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": "(c++) v8::internal::Map::RawCopy",
      "line": undefined,
      "name": "(c++) v8::internal::Map::RawCopy",
      "selfWeight": 22888,
      "totalWeight": 22888,
    },
    Frame {
      "col": NaN,
      "file": "module.js",
      "key": "Module.runMain module.js:663:26",
      "line": 663,
      "name": "Module.runMain",
      "selfWeight": 0,
      "totalWeight": 64151,
    },
    Frame {
      "col": NaN,
      "file": "module.js",
      "key": "Module._load module.js:443:24",
      "line": 443,
      "name": "Module._load",
      "selfWeight": 0,
      "totalWeight": 64151,
    },
    Frame {
      "col": NaN,
      "file": "module.js",
      "key": "tryModuleLoad module.js:505:23",
      "line": 505,
      "name": "tryModuleLoad",
      "selfWeight": 0,
      "totalWeight": 64151,
    },
    Frame {
      "col": NaN,
      "file": "module.js",
      "key": "Module.load module.js:536:33",
      "line": 536,
      "name": "Module.load",
      "selfWeight": 0,
      "totalWeight": 64151,
    },
    Frame {
      "col": NaN,
      "file": "module.js",
      "key": "Module._extensions..js module.js:633:37",
      "line": 633,
      "name": "Module._extensions..js",
      "selfWeight": 0,
      "totalWeight": 64151,
    },
    Frame {
      "col": NaN,
      "file": "module.js",
      "key": "Module._compile module.js:581:37",
      "line": 581,
      "name": "Module._compile",
      "selfWeight": 0,
      "totalWeight": 64151,
    },
    Frame {
      "col": NaN,
      "file": "/Users/jlfwong/code/speedscope/sample/programs/javascript/simple.js",
      "key": " /Users/jlfwong/code/speedscope/sample/programs/javascript/simple.js:1:11",
      "line": 1,
      "name": "(anonymous)",
      "selfWeight": 0,
      "totalWeight": 64151,
    },
    Frame {
      "col": NaN,
      "file": "/Users/jlfwong/code/speedscope/sample/programs/javascript/simple.js",
      "key": "a /Users/jlfwong/code/speedscope/sample/programs/javascript/simple.js:1:73",
      "line": 1,
      "name": "a",
      "selfWeight": 0,
      "totalWeight": 64151,
    },
    Frame {
      "col": NaN,
      "file": "/Users/jlfwong/code/speedscope/sample/programs/javascript/simple.js",
      "key": "b /Users/jlfwong/code/speedscope/sample/programs/javascript/simple.js:8:11",
      "line": 8,
      "name": "b",
      "selfWeight": 10044,
      "totalWeight": 53997,
    },
    Frame {
      "col": NaN,
      "file": "/Users/jlfwong/code/speedscope/sample/programs/javascript/simple.js",
      "key": "d /Users/jlfwong/code/speedscope/sample/programs/javascript/simple.js:20:11",
      "line": 20,
      "name": "d",
      "selfWeight": 51562,
      "totalWeight": 51562,
    },
    Frame {
      "col": NaN,
      "file": "/Users/jlfwong/code/speedscope/sample/programs/javascript/simple.js",
      "key": "c /Users/jlfwong/code/speedscope/sample/programs/javascript/simple.js:14:11",
      "line": 14,
      "name": "c",
      "selfWeight": 2545,
      "totalWeight": 10154,
    },
  ],
  "stacks": Array [
    "(anonymous);startup;setupGlobalVariables;NativeModule.require;NativeModule.compile;(anonymous);NativeModule.require;NativeModule.compile;(anonymous);(c++) v8::internal::Runtime_CreateArrayLiteral;(c++) v8::internal::JSFunction::EnsureHasInitialMap 29.38ms",
    "(anonymous);startup;setupGlobalConsole;setupInspectorCommandLineAPI;NativeModule.require;NativeModule.compile;(anonymous);NativeModule.require;NativeModule.compile;(anonymous);(c++) v8::internal::Runtime_StoreIC_Miss;(c++) v8::internal::Map::RawCopy 22.89ms",
    "(anonymous);startup;Module.runMain;Module._load;tryModuleLoad;Module.load;Module._extensions..js;Module._compile;(anonymous);a;b;d 37.52ms",
    "(anonymous);startup;Module.runMain;Module._load;tryModuleLoad;Module.load;Module._extensions..js;Module._compile;(anonymous);a;c;d 1.28ms",
    "(anonymous);startup;Module.runMain;Module._load;tryModuleLoad;Module.load;Module._extensions..js;Module._compile;(anonymous);a;b;d 2.55ms",
    "(anonymous);startup;Module.runMain;Module._load;tryModuleLoad;Module.load;Module._extensions..js;Module._compile;(anonymous);a;c;d 1.27ms",
    "(anonymous);startup;Module.runMain;Module._load;tryModuleLoad;Module.load;Module._extensions..js;Module._compile;(anonymous);a;b;d 1.28ms",
    "(anonymous);startup;Module.runMain;Module._load;tryModuleLoad;Module.load;Module._extensions..js;Module._compile;(anonymous);a;c;d 1.27ms",
    "(anonymous);startup;Module.runMain;Module._load;tryModuleLoad;Module.load;Module._extensions..js;Module._compile;(anonymous);a;b;d 1.29ms",
    "(anonymous);startup;Module.runMain;Module._load;tryModuleLoad;Module.load;Module._extensions..js;Module._compile;(anonymous);a;c;d 1.26ms",
    "(anonymous);startup;Module.runMain;Module._load;tryModuleLoad;Module.load;Module._extensions..js;Module._compile;(anonymous);a;b;d 1.31ms",
    "(anonymous);startup;Module.runMain;Module._load;tryModuleLoad;Module.load;Module._extensions..js;Module._compile;(anonymous);a;c;d 2.52ms",
    "(anonymous);startup;Module.runMain;Module._load;tryModuleLoad;Module.load;Module._extensions..js;Module._compile;(anonymous);a;c 1.27ms",
    "(anonymous);startup;Module.runMain;Module._load;tryModuleLoad;Module.load;Module._extensions..js;Module._compile;(anonymous);a;b 3.62ms",
    "(anonymous);startup;Module.runMain;Module._load;tryModuleLoad;Module.load;Module._extensions..js;Module._compile;(anonymous);a;c 1.28ms",
    "(anonymous);startup;Module.runMain;Module._load;tryModuleLoad;Module.load;Module._extensions..js;Module._compile;(anonymous);a;b 6.42ms",
  ],
}
`;
